runtime/compiler/x/codegen/J9TreeEvaluator.cpp

```diff
@@ -7908,27 +7908,47 @@ J9::X86::TreeEvaluator::VMarrayStoreCHKEvaluator(
       // -------------------------------------------------------------------------
       //
       // If the component type is java.lang.Object then the store always succeeds.
+      // On AOT, only do this when using the SVM; otherwise, it's safer to just
+      // skip this check than 1) incorrectly relocate the java/lang/Object class
+      // or 2) crash while trying to relocate.
       //
       // -------------------------------------------------------------------------
+      if (!cg->needClassAndMethodPointerRelocations() || comp->getOption(TR_UseSymbolValidationManager))
+         {
+         TR_OpaqueClassBlock *objectClass = fej9->getSystemClassFromClassName("java/lang/Object", 16);
+         objectClass = fej9->getArrayClassFromComponentClass(objectClass);
 
-      TR_OpaqueClassBlock *objectClass = fej9->getSystemClassFromClassName("java/lang/Object", 16);
-      objectClass = fej9->getArrayClassFromComponentClass(objectClass);
+         if (TR::Compiler->target.is64Bit())
+            {
+            TR_ASSERT((((uintptr_t)objectClass) >> 32) == 0, "TR_OpaqueClassBlock must fit on 32 bits when using class pointer compression");
 
-      if (TR::Compiler->target.is64Bit())
-         {
-         TR_ASSERT((((uintptr_t)objectClass) >> 32) == 0, "TR_OpaqueClassBlock must fit on 32 bits when using class pointer compression");
-         instr = generateRegImmInstruction(CMP4RegImm4, node, destComponentClassReg, (uint32_t) ((uint64_t) objectClass), cg);
-         }
-      else
-         {
-         instr = generateRegImmInstruction(CMP4RegImm4, node, destComponentClassReg, (int32_t)(uintptr_t) objectClass, cg);
-         }
+            if (cg->needClassAndMethodPointerRelocations())
+               {
+               TR::Register *objectClassReg = scratchRegisterManager->findOrCreateScratchRegister();
+               // MOV8RegImm64 because the relocation infrastructure can't currently relocate a 32-bit pointer
+               instr = generateRegImm64Instruction(MOV8RegImm64, node, objectClassReg, (uintptr_t) objectClass, cg, TR_ClassPointer);
+               generateRegRegInstruction(CMP4RegReg, node, destComponentClassReg, objectClassReg, cg);
+               scratchRegisterManager->reclaimScratchRegister(objectClassReg);
```

Comment:
Alternatively, we can store `objectClass` in a data snippet and use `CMP4RegMem` so that the scratch register can be avoided.